### ISSUE TITLE
[BUG]: Incorrect screen time

### ISSUE THREAD

**Issue by OndrejKulhavy** [tanujnotes/Olauncher#545]
I've noticed that my screen time is unusually high. On 00:01 midnight, I have over an hour of screen time, which doesn't seem logical.

Here's an example from just now: Digital Wellbeing is displaying 55 minutes, which seems realistic, while OLauncher is indicating 3 hours and 12 minutes.

<div>
  <img src="https://github.com/user-attachments/assets/e07328f7-57b6-4490-9788-f64f3605d646" alt="Image 1" width="300">
  <img src="https://github.com/user-attachments/assets/3354c305-0421-497f-b63a-1303ac0c77f6" alt="Image 2" width="300">
</div>

**Comment by HeCodes2Much** [comment]
![Screenshot_20241121-000449.png](https://github.com/user-attachments/assets/50c4b155-ebd2-4502-9512-a8708b972432)

Ya mine says 7 hours when it's just past midnight.

I think its calculating the past 24 hours not resetting at midnight.

**Comment by OndrejKulhavy** [comment]
That's stupid, I guess you are right!

**Comment by daaanny90** [comment]
I have the same problem. Samsung galaxy S23 with Android 14

**Comment by OndrejKulhavy** [comment]
To be honest, since it probably only counts the last 24 hours, it's a somewhat useless indicator. I'm spending time on my phone, yet the time displayed is either stagnant or decreasing. 💀

**Comment by eddex** [comment]
https://github.com/tanujnotes/Olauncher/blob/fbcc721e9c97368c5b621270cc4e183920febea2/app/src/main/java/app/olauncher/MainViewModel.kt#L255-L273

It seems like the intention was to show the screen time from the start of the current day until 24h after. Which seems correct. But there must be a bug somewhere. Might also be time zone related, but that's just a wild guess.

**Comment by HeCodes2Much** [comment]
> https://github.com/tanujnotes/Olauncher/blob/fbcc721e9c97368c5b621270cc4e183920febea2/app/src/main/java/app/olauncher/MainViewModel.kt#L255-L273
> 
> It seems like the intention was to show the screen time from the start of the current day until 24h after. Which seems correct. But there must be a bug somewhere. Might also be time zone related, but that's just a wild guess.

Ya that is how it is meant to work but some reason it doesn't take the current time even changing the 0 for hour to 9 for 9am it still displays the same data so the hour is not affecting anything. I have been playing with that part of the code for 3 days now and can't figure out why the hour isn't effected

**Comment by tanujnotes** [comment]
It has something to do with the UTC timezone but I haven't been able to figure out a solution. The response Android provides for the screen time is all over the place. I tried different screen time apps and they all show different values, although not as wrong as Olauncher. Working on it.

**Comment by HeCodes2Much** [comment]
> It has something to do with the UTC timezone but I haven't been able to figure out a solution. The response Android provides for the screen time is all over the place. I tried different screen time apps and they all show different values, although not as wrong as Olauncher. Working on it.

Take a look at [usageDirect](https://f-droid.org/en/packages/godau.fynn.usagedirect/) they seem to do a pretty good job with it.

**Comment by Evertvz** [comment]
Hello, 
Having the same issue here on a fairphone.

**Comment by CreativeCodeCat** [comment]
> Hello, 
> Having the same issue here on a fairphone.

FairPhone 5 Crew :)

**Comment by CreativeCodeCat** [comment]
@ArchUsr64 sadly that doesn't fix the issue at all.

**Comment by ArchUsr64** [comment]
> [@ArchUsr64](https://github.com/ArchUsr64) sadly that doesn't fix the issue at all.

Yeah, it's very strange. It worked on my phone for a few days and then broke again. Likely an issue with the underlying android API.

**Comment by CreativeCodeCat** [comment]
ya i have spent hours trying to work out why it is not working and even tried to look at how [usagedirect](https://f-droid.org/en/packages/godau.fynn.usagedirect/) does it but no luck sadly :(

**Comment by OndrejKulhavy** [comment]
I found a fix by filtering today’s values

**Comment by CreativeCodeCat** [comment]
> However, it also counts the time spent in the launcher, so the time differs a little bit compared to the Digital Wellbeing time.

I tried this along with a blacklist of apps using the code below but it still shows 2 hours where as Digital Wellbeing is showing 3 mins and the old way shows 12 mins

```
private fun parseBlacklistXML(context: Context): List<String> {
        val packageNames = mutableListOf<String>()

        // Obtain an XmlPullParser for the blacklist.xml file
        context.resources.getXml(R.xml.blacklist).use { parser ->
            while (parser.eventType != XmlPullParser.END_DOCUMENT) {
                if (parser.eventType == XmlPullParser.START_TAG && parser.name == "app") {
                    val packageName = parser.getAttributeValue(null, "packageName")
                    packageNames.add(packageName)
                }
                parser.next()
            }
        }

        return packageNames
    }
```


Your way

![Image](https://github.com/user-attachments/assets/8064bef6-3bd9-4d37-af45-0e556eb1df2e)

Old Way

![Image](https://github.com/user-attachments/assets/ad58ee0c-7c0b-450e-abd8-b443706e4fd4)

**Comment by OndrejKulhavy** [comment]
Interesting. After filtering out the launcher app, I have a difference of ±5 minutes. I will test the behavior after midnight.

**Comment by CreativeCodeCat** [comment]
I will give it a test tomorrow too and see if them new changes fix it.

**Comment by OndrejKulhavy** [comment]
For reference

<div style="display: flex; gap: 10px; align-items: center;">
  <div>
    <strong>Original version:</strong><br>
    <img src="https://github.com/user-attachments/assets/01b699d8-661f-44c4-83c8-5be4970b12e7" width="300"/>
  </div>
  <div>
    <strong>After My Changes:</strong><br>
    <img src="https://github.com/user-attachments/assets/6eb391df-e020-4274-8e96-e6dfe6705c80" width="300"/>
  </div>
  <div>
    <strong>Reference (Digital Wellbeing):</strong><br>
    <img src="https://github.com/user-attachments/assets/600acdc9-d9d7-4072-af5b-752632d5fd16" width="300"/>
  </div>
</div>

**Comment by tanujnotes** [comment]
Try this: https://github.com/tanujnotes/Olauncher/releases/tag/v5.2.28

**Comment by OndrejKulhavy** [comment]
Interesting. Can you please share your weekly wellbeing stats?

**Comment by Npa972** [comment]
![Image](https://github.com/user-attachments/assets/492fd138-b31d-4821-975b-422e502caebb)

**Comment by tanujnotes** [comment]
The new screen time calculation method seems to be failing on Samsung devices. Fixing it.

**Comment by Npa972** [comment]
Yesterday at midnight time screen reset to 0.
This morning time screen seems OK.
I don't understand, It's beyond my skills.
There is a difference of a few minutes compared to well being calculation.
Thank you.

![Image](https://github.com/user-attachments/assets/b5454fae-8adc-4329-acbf-e76df404a3f4)

**Comment by nmariano9** [comment]
Just chiming in to say that there is still a bug for my screen time display on a Google Pixel 8.

**Comment by ConnorMooneyhan** [comment]
This is still an issue for me as well. I'm on a Samsung Galaxy S25 Ultra.

**Comment by adrianjost** [comment]
It started to become an issue in the last weeks for me too.. I now often wake up with Screen Time values way higher than the hours of the day elapsed (with actual 0 min). I also reach 25+hours of screen time.
A couple weeks ago, the values were identical to the Digital Well Being App on my S24.

**Comment by RyanGaudion** [comment]
Same issue here: Huawei P30 Pro
At 8:05am - OLauncher is showing 8h 52m 
Actual = 7 mins
Last 7 Days: 7 min, 56 min, 1h 10m, 2h 21m, 3h 24m, 56m, 3h 52m

**Comment by kruinkriebel** [comment]
Also experiencing this issue on Pixel 6. 

Would be nice if there was an option to just hide the amount, if it's not functioning... It would also be nice to be able to blacklist some apps, like navigation. 

I guess there is no API inside Android to request this value, instead of calculating it yourself?

**Comment by Eleusynia** [comment]
Not sure if this has already been spotted but if you don't switch your device off, the counter will reset at midnight. It's not a fix, but I guess if you don't want the timer to be counting the period from when your device was switched off before 12am to the time you wake up, this could be a short term workaround.

**Comment by maoryouavian** [comment]
Any update on this issue?

**Comment by m-ad** [comment]
I still have this issue with v6.7.1. Olauncher shows "8h 39m". Clicking on the number opens the Digital Wellbeing app and there it says "2h 2min" instead for today. Yesterday was "2h 32min", so Olauncher shows more than the last 24 hours. Phone is a motorola edge 20 lite (XT2139-1) running Android 13, timezone UTC+2.

**Comment by tanujnotes** [comment]
Occasionally the screen time will be wrong. I'm not entirely sure why but the leading theory is that sometimes Olauncher calculates something running in the background as a foreground app, resulting in absurdly large screen time. Most of the time it corrects itself in a day or two. If I can get any hint about exactly what app or background process is causing this, I might be able to replicate it on my device and fix it.
